### ISSUE TITLE
FieldBuilder ignores nullable argument when it is passed along with type

### ISSUE THREAD

**Issue by vkukarau** [graphql-dotnet/graphql-dotnet#1290]
## Description
FieldBuilder ignores `nullable` argument if `type` is specified.

### Steps to reproduce
Define GraphQL type with fields as follows:
```
    public class AppointmentType : ObjectGraphType<Appointment>
    {
        public AppointmentType()
        {
            Field(x => x.Date, false, typeof(DateTimeGraphType));
            Field(x => x.Status, false, typeof(EnumerationGraphType<AppointmentStatus>));
            Field(x => x.Description, false, typeof(StringGraphType));
            Field(x => x.Title, false);
       }
    }   
```
Note that all fields are set as non-nullable, first 3 fields have `type` argument, and the last one doesn't have it.

### Expected result
Fields in generated GraphQL schema should look like:
```
date: DateTime!
status: AppointmentStatus!
description: String!
title: String!
```

### Actual result
Fields in generated GraphQL schema look like:
```
date: DateTime
status: AppointmentStatus
description: String
title: String!
```
Note that all fields are specified as non-nullable, but only `title` is properly defined in the resulting schema

### Environment
nuget GraphQL 2.4.0
Windows

**Comment by joemcbride** [comment]
This is expected behavior.  If you are choosing to override the GraphType, that means you have full control over the exact definition.

**Comment by MisterJimson** [comment]
While that makes sense, it's not a great developer experience to explicitly pass `false` to a parameter called 'nullable' and have the result not reflect that.

Do you think a reasonable change could be to wrap the type in a NonNullGraphType if `false` is set here to avoid this pitfall?

**Comment by joemcbride** [comment]
`nullable` is an optional parameter to that method.

https://github.com/graphql-dotnet/graphql-dotnet/blob/444cc529a15526aa9b70376ec54bbe5b6d33fdcb/src/GraphQL/Types/ComplexGraphType.cs#L278-L282

What do you tell the person who then does this?

```csharp
Field(x => x.Description, type: typeof(NonNullGraphType<StringGraphType>));
```

And now ends up with:

```csharp
description: String!!
```

This would be a breaking behavior change.  Someone has to change their code.  Either you or them.

We could update the API so there are two methods, one that only accepts the type and one that only accepts the nullable, so there is never the combination.

**Comment by MisterJimson** [comment]
Well if the type is already a `NonNullGraphType`, we wouldn't need to wrap it in a `NonNullGraphType` when nullable is false.

Completely understand the backwards compatibility concern though. Perhaps something to think about for 3.0 then?

I am fine updating my code and already have. This issue was just raised as it seemed a like a pretty clear bug that wasn't in the issue tracker ("I'm telling this field it can't be null but it is!").

**Comment by sungam3r** [comment]
When I read the `Field<TProperty>` source code, this moment also confused me with the nullable argument. I decided to deal with this issue later, as I intuitively understood that the changes could be backward incompatible and change the design a little.

**Comment by sungam3r** [comment]
I myself came across a misunderstanding of this API. My work colleagues are similar. So I think we can split this method into two.

**Comment by sungam3r** [comment]
Relates to #1567

**Comment by Shane32** [comment]
@sungam3r Shall we close this or do you have a suggestion here?  From the original issue, I would say it's as-designed.  Now we have field builders which can infer the graph type from the .NET type also.

**Comment by sungam3r** [comment]
I wanted to redesign it in some way or another and still want. No concrete suggestions yet but I think that after field builders redesign I can return to this issue for v8 milestone. I can look into this one tonight.

**Comment by sungam3r** [comment]
Current options for object graph type (I consider #3583 as a base):

1. Field("field", new StringGraphType()); // OK - name+type
2. Field("field", typeof(StringGraphType)); // OK - name+type
3. Field<StringGraphType>("field"); // OK - name+type
4. Field<string>("field", true /*opt*/); // NOT OK - name+type+modifier
5. Field(x => x.OrderId, true /*opt*/, typeof(StringGraphType) /*opt*/); // NOT OK - name+2 types+modifier
6. Field("field", x => x.OrderId, true /*opt*/, typeof(StringGraphType) /*opt*/); // NOT OK - name+2 types+modifier
7. Field<StringGraphType, string>("field"); // OK - name+type

What I would like to do - rewrite API to get rid of 4,5,6 options.

I suggest to consider such new APIs for `ObjectFieldBuilder` (as well as for other builders):
```cs
        public virtual ObjectFieldBuilder<TSourceType, TReturnType> Nullable(bool nullable = true)
        {
            /* Code to wrap or unwrap FieldType.ResolvedType / FieldType.Type */
            return this;
        }

        public virtual ObjectFieldBuilder<TSourceType, TReturnType> NotNull() => Nullable(false);
```

Usage:

```cs
 Field(x => x.OrderId).NonNull();
 Field(x => x.OrderId).Nullable();
```

Of course new methods may be applied after any `Field()` overload, not only one with expression. `NonNull` and `Nullable` methods will be idempotent, i.e. `Field(x => x.OrderId).NonNull().NonNull().NonNull().NonNull().NonNull()` will not throw as well as `Field(x => x.OrderId).Nullable().Nullable().Nullable().Nullable().Nullable().Nullable()`.

As a result API surface will be reduced on graph types but increased on builders. I see such design as allowing to have a more straighforward developer experience.

Side note: I would like to get rid of option 6 even without taking into account all duscussions here since
`Field("field", x => x.OrderId)` can be just changed to `Field(x => x.OrderId).Name("field")`.

My whole idea for fields API - make it as minimalistic as possible (4 overloads maybe?) on GraphType classes moving all additional configuration as fluent API in builders.

**Comment by Shane32** [comment]
It is important to distinguish the use case -- for output fields vs inputs fields.  Input fields require an `Expression` in order to map the property.  Output fields should allow them as a shortcut, but often will not, as they will use `.Resolve`.  There are also interface fields which will not have a resolver.  For purposes of discussion here we can lump them with output fields that use a `.Resolve` syntax.

Second, my opinion is that we should maintain the ability to map simple properties quickly and easily.  For example, we should continue to support `Field("Company", x => x.CompanyName, true)` and not require `Field<StringGraphType>().Name("Company").Resolve(x => x.Source.CompanyName)` or some such complicated thing.  The field builders should be used primarily for (a) any resolver more complicated than a simple property mapping, and (b) metadata.  Field builders can also be used in any advanced scenario.

Also, when we rewrote many of the field builders for v7, I expressed that we should maintain most of the old methods (with obsolete tags) for a couple years before removing from the codebase.  I feel this is very important.

### Input types

Now let's look at what we have now from a 'input type' view: only options 5 and 6 apply.  They are really both the same, with the name optional on the front end.  I certainly would not remove them -- they are the only two methods useful for input types -- although perhaps we could split them into two methods.  For instance:

a. `Field(x => x.OrderId, true);`
b. `Field(x => x.OrderId, typeof(IdGraphType));`

I think it's important to preserve this functionality.  As to the suggestion here, making `false` add a non-null wrapper around the specified graph type, that does not currently fit with any other functionality we have, and would be a non-obvious breaking change.  I would leave as-is, although I agree that may have been a better design when this project started.

We might also ask at this point: do any of the other overloads apply for input types?  The answer is yes: when `ParseDictionary` is in use, then the expression is ignored and only the name and graph type of each field is relevant.  So, we can eliminate option 7 for input types only, as the .NET type is meaningless in that situation.

### Output types

Now from an 'output types' point of view, I pretty much have the same opinion on options 5 & 6.  There are many many many cases that a user has some code like this:

```cs
Field(x => x.Id, false, typeof(NonNullGraphType<IdGraphType>)); // specific type
Field(x => x.Name); // non-null
Field("Company", x => x.CompanyName, true); // nullable
Field(x => x.Address, true); //nullable
Field(x => x.City, true); //nullable
Field(x => x.State, true); //nullable
```

I think (a) the name should be the first argument, except when a property is specified and name is implicit, (b) name should be able to be overridden, and (c) must be able to override the graph type.  The graph type is often overridden for id and date fields.  The existing design allows for this, and as can be seen above, requires "options 5 & 6".  To remove options 5/6 would require something like this:

```cs
Field<NonNullGraphType<IdGraphType>>("Id").Resolve(x => x.Source.Id); // specific type
Field<string>("Name").Resolve(x => x.Source.Name); // non-null
Field<string>("Company", true).Resolve(x => x.Source.CompanyName); // nullable
Field<string>("Address", true).Resolve(x => x.Source.Address, true); //nullable
Field<string>("City", true).Resolve(x => x.Source.City); //nullable
Field<string>("State", true).Resolve(x => x.Source.State, true); //nullable
```

The above would be such a disservice to our users, as it would be much harder to read/write.

### Eliminating option 4

You also mentioned eliminating option 4.  The issue is that NRT attribute for the `string?` or `string` generic type is never preserved in that scenario.  There is no way to infer the proper graph type without the boolean argument.

Finally, you mentioned adding `.Nullable` and `.NonNull`.  This has some merit, as the syntax is readable.  To reiterate your sample, it would allow:

```cs
//old
Field(x => x.OrderId, true);

//new
Field(x => x.OrderId).Nullable();
```

It would be required to assume non-null or else there would be subtle breaking changes, which I don't think we can afford here.  Also I wouldn't eliminate the prior syntax, as it is heavily used and works well as-is.

Keep in mind that this only works for the outermost level of nullability (as before) -- it doesn't help with the inner type of list types.

### Conclusion

- I don't think any of the 7 methods should be eliminated.  I think they all work well.
- For input type field builders, option 7 should be eliminated.
- I would add `.Nullable()` and `.NonNull()` as additional methods (like `.Name()`) that are redundant but fit well
- I would not remove any `Field` methods currently marked as `[Obsolete]` for v8 for output field types, but I would eliminate them for input and interface field types.

**Comment by Shane32** [comment]
> My whole idea for fields API - make it as minimalistic as possible (4 overloads maybe?) on GraphType classes moving all additional configuration as fluent API in builders.

I agree in concept, but I think it should as a 'minimal' syntax define the following:
1. Name
2. Graph type
3. .NET type for output types
4. Property, when possible (optional)

Since `Field(x => x.Name)` defines all 4 of these points, it's a great API.  But sometimes you need `Field<TimeSpanSecondsGraphType, TimeSpan>("Notes")` when the resolver is more complex than a simple property accessor.

So if anything, we could eliminate (mark as obsolete) overloads that miss one of the first three points above.  So for output types, this would be marking `Field<IdGraphType>("Id")` as obsolete because it does not define that the .NET type is `string`.  That would be overload options 1, 2, and 3 from your list above (making sure we have a suitable replacement for option 1 that is similar to option 7).  But for input types this is fine because input types should not use a .NET type for the field builder.

**Comment by Shane32** [comment]
@gao-artur You might be interested in the latter half of this old thread, if you've not seen it before.

**Comment by gao-artur** [comment]
I don't like catchy APIs that require reading documentation to use them properly. So, for v8, I suggest doing one of these
1. Split the `nullable` and `type` to separate methods
2. Make them work together

While I can see the value in making them work together, I think using 2 parameters when it's possible to use one is not the best practice.
For example this:
```c#
Field(x => x.Description, false, typeof(StringGraphType));
```
Is shorter than this:
```c#
Field(x => x.Description, typeof(NonNullGraphType<StringGraphType>));
```

But this combination will raise questions:
```c#
Field(x => x.Description, false, typeof(NonNullGraphType<StringGraphType>));
```
Some users may assume the additional `nullable: false` will be ignored, while others may wonder if it produces an invalid `string!!`. Then, they will need to read the docs or try and see the final result.

So, my vote goes to 1: Remove the ambiguity.

In v7, we can add the automatic code fix to remove the `nullable` usage when it is used with `type` together. It should be safe because this argument is ignored anyway.
```c#
// old
Field(x => x.Description, false, typeof(NonNullGraphType<StringGraphType>));

//new
Field(x => x.Description, type: typeof(NonNullGraphType<StringGraphType>));
```

The automatic code fix is only possible when a `true/false/null` constant is used for `nullable`. This scenario is more complex but should be rare:
```c#
var nullable = GetNullable(...);
var type = GetType(...);
Field(x => x.Description, nullable, type);
```

I can rewrite is like this
```c#
var nullable = GetNullable(...);
var type = GetType(...);

if (type != null)
    Field(x => x.Description, type: type);
else if (nullable != null)
    Field(x => x.Description, nullable);
else
    Field(x => x.Description);
```

But since it should be a rare case, I'd prefer to leave it to user to fix manually.

Then, in v8, we can split the method and add a comment in the upgrade notes to upgrade first to the latest v7 and use all the automatic code fixes for easier migration.

**Comment by Shane32** [comment]
I’m all for your PR where there’s two separate methods. I just would mark the old one obsolete, that’s all.

**Comment by gao-artur** [comment]
I thought there would be ambiguity when defining these overloads, but they seem to work fine (pseudo code):
```c#
[Obsolete]
public FieldBuilder Field(Expression expression, bool? nullable = null, Type type = null);

public FieldBuilder Field(Expression expression, bool nullable);

public FieldBuilder Field(Expression expression, Type type);

public FieldBuilder Field(Expression expression);
```

The obsolete is intended to allow users to gradually prepare for breaking change. But an automatic code fix makes it easy, so I don't see a real value in keeping the obsolete overload. But since these overloads have no ambiguity, we can keep the obsolete one if you prefer.

**Comment by gao-artur** [comment]
The method was separated in #3971 and an automatic code fix for migration was prepared in #3973. I think we can close this issue.

**Comment by Shane32** [comment]
Agreed
